composer/datasets/synthetic_lm.py

```diff
@@ -0,0 +1,231 @@
+            if using pip or `conda install -c conda-forge transformers tokenizers` if using Anaconda.""")) from e
+
+    # generate a synthetic dataset with reasonable defaults is none is provided
+    if dataset is None:
+        num_samples = 100
+        chars_per_sample = 128
+        column_names = ['sentence']
+        dataset = SyntheticHFDataset(num_samples=num_samples,
+                                     chars_per_sample=chars_per_sample,
+                                     column_names=column_names).generate_dataset()
+
+    # change a columnar dataset into a list
+    flattened_columns = [dataset[key] for key in dataset.column_names if key != 'idx']
+    # flatten the list of lists into a single list
+    flattened_dataset = []
+    for sublist in flattened_columns:
+        for item in sublist:
+            flattened_dataset.append(item)
+
+    if "bert" in tokenizer_family:
+        tokenizer_params = _generate_bert_tokenizer_params(flattened_dataset)
+    elif "gpt2" in tokenizer_family:
+        tokenizer_params = _generate_gpt2_tokenizer_params()
+    else:
+        raise ValueError(f"Synthetic tokenizers for tokenizer family {tokenizer_family} are currently unsupported.")
+
+    tokenizer = Tokenizer(tokenizer_params.tokenizer_model)
+    tokenizer.enable_padding(direction="right",
+                             pad_id=0,
+                             pad_type_id=0,
+                             pad_token=tokenizer_params.pad_token,
+                             pad_to_multiple_of=8)
+    tokenizer.normalizer = tokenizer_params.normalizer  # type: ignore
+    tokenizer.pre_tokenizer = tokenizer_params.pre_tokenizer  # type: ignore
+    tokenizer.decoder = tokenizer_params.decoder  # type: ignore
+    tokenizer_trainer = tokenizer_params.trainer_cls(
+        vocab_size=vocab_size,
+        initial_alphabet=tokenizer_params.initial_alphabet,
+        special_tokens=tokenizer_params.special_tokens,
+    )
+    tokenizer.train_from_iterator(flattened_dataset, trainer=tokenizer_trainer)
+
+    # save the tokenizer config
+    tmp_tokenizer_dir = mkdtemp()
+    tmp_tokenizer_file = join(tmp_tokenizer_dir, "tokenizer.json")
+    tokenizer.save(tmp_tokenizer_file)
+
+    # save the vocabulary and potential merges file
+    tokenizer_params.tokenizer_model.save(tmp_tokenizer_dir)  # type: ignore
+
+    # the .from_pretrained method doesn't load our padding for some reason, so we save it as a special kwarg
+    tmp_tokenizer_config = join(tmp_tokenizer_dir, "tokenizer_config.json")
+    with open(tmp_tokenizer_config, "w") as f:
+        json.dump({"pad_token": tokenizer_params.pad_token}, f)
+
+    # instantiate the new tokenizer
+    if not issubclass(tokenizer_params.tokenizer_cls, PreTrainedTokenizer):
+        raise ValueError(f"{tokenizer_params.tokenizer_cls} should sub-class transformers.PreTrainedTokenizer.")
+    tokenizer = tokenizer_params.tokenizer_cls.from_pretrained(tmp_tokenizer_dir)
+
```

Comment:
should the tempdir be cleaned up here? if so, perhaps instead should use the `TemporaryFolder` context instead of the low level `mkdtemp`

Reply:
The temporary directory should not be cleaned up, because the tokenizer folder that is created needs to exist for the model to access the tokenizer as well.

This might imply a potential race condition if the temporary directory is cleaned up by the OS before the model gets instantiated? Is that a concern?

Reply:
Discussed offline; the folder will be passed in as an argument into the function.

Reply:
Done!